### PR TITLE
Export plugin plots

### PR DESCRIPTION
**Pull request by cshanahan1**
Add ability to export plots from various plugins through export plugin.

@kecnry did much of this, and the changes I made on top of his are:

1. Filtering out empty plots (since some plots, like aperture photometry, exist but are empty) so they're not displayed in the plugin.
2. Test coverage

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/app.py [hunk 1]
```diff
@@ -381,6 +384,14 @@ def _on_plugin_table_added(self, msg):
         key = f"{msg.plugin._plugin_name}:{msg.table._table_name}"
         self._plugin_tables.setdefault(key, msg.table.user_api)
 
+    def _on_plugin_plot_added(self, msg):
+        if msg.plugin._plugin_name is None:
+            # plugin was instantiated after the app was created, ignore
+            return
+        # TODO: make sure plot._plot_name is set
```
Comment (kecnry) [inline jdaviz/app.py:391]:
```suggestion
```

File: jdaviz/core/template_mixin.py [hunk 2]
```diff
@@ -4358,9 +4479,16 @@ def __init__(self, plugin, viewer_type='scatter', app=None, *args, **kwargs):
                     ]
         self._initialize_toolbar()
 
+        plugin.session.hub.broadcast(PluginPlotAddedMessage(sender=self))
+        plugin.session.hub.broadcast(PluginPlotModifiedMessage(sender=self))
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:4483]: Is there a reason to broadcast both here?

```suggestion
```

File: jdaviz/core/template_mixin.py [hunk 3]
```diff
@@ -4482,9 +4620,13 @@ def clear_marks(self, *mark_labels):
                     else:
                         mark.x, mark.y = [], []
 
+        self._plugin.session.hub.broadcast(PluginPlotModifiedMessage(sender=self))
+
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:4624]: I don't think anything is changed when just accessing the marks

```suggestion
```

File: jdaviz/core/template_mixin.py [hunk 4]
```diff
@@ -4520,19 +4665,25 @@ def add_line(self, label, x=[], y=[], xnorm=False, ynorm=False, **kwargs):
                               colors=kwargs.pop('color', kwargs.pop('colors', 'gray')),
                               **kwargs)
 
+        self._plugin.session.hub.broadcast(PluginPlotModifiedMessage(sender=self))
+
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:4669]: Should be covered already by the broadcast in `_add_mark`

```suggestion
```

File: jdaviz/core/template_mixin.py [hunk 5]
```diff
@@ -4520,19 +4665,25 @@ def add_line(self, label, x=[], y=[], xnorm=False, ynorm=False, **kwargs):
...
     def add_scatter(self, label, x=[], y=[], xnorm=False, ynorm=False, **kwargs):
         return self._add_mark(bqplot.Scatter, label, x=x, y=y,
                               xnorm=xnorm, ynorm=ynorm,
                               colors=kwargs.pop('color', kwargs.pop('colors', 'gray')),
                               **kwargs)
 
+        self._plugin.session.hub.broadcast(PluginPlotModifiedMessage(sender=self))
+
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:4677]: Should be covered already by the broadcast in `_add_mark`

```suggestion
```

File: jdaviz/core/template_mixin.py [hunk 6]
```diff
@@ -4520,19 +4665,25 @@ def add_line(self, label, x=[], y=[], xnorm=False, ynorm=False, **kwargs):
...
     def add_bins(self, label, sample=[0], bins=2, density=True, **kwargs):
         # NOTE: initializing with bins=1 breaks the figure until a resize event
         return self._add_mark(bqplot.Bins, label, sample=sample, bins=bins,
                               density=density,
                               colors=kwargs.pop('color', kwargs.pop('colors', 'gray')),
                               **kwargs)
 
+        self._plugin.session.hub.broadcast(PluginPlotModifiedMessage(sender=self))
+
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py:4686]: Should be covered already by the broadcast in `_add_mark`

```suggestion
```

File: jdaviz/core/template_mixin.py (hunk 2 above)
Comment (cshanahan1) [reply jdaviz/core/template_mixin.py:4483]: i cant remember why i did this, but there was a reason at the time haha. ill remove and test without.

Comment (kecnry) [reply jdaviz/core/template_mixin.py:4483]: If it turns out to be necessary, please just add a comment so we don't forget and accidentally remove it in the future

File: jdaviz/core/template_mixin.py (hunk 3 above)
Comment (cshanahan1) [reply jdaviz/core/template_mixin.py:4623]: i thought clearing marks would count as modifying the plot?

Comment (kecnry) [reply jdaviz/core/template_mixin.py:4624]: ah, I misread the diff and thought it was in the marks property.  Yes, let's keep this one gut remove the one in `clear_all_marks` then (which calls this)

File: jdaviz/core/template_mixin.py (hunk 2 above)
Comment (cshanahan1) [reply jdaviz/core/template_mixin.py:4483]: i tested and i don't think its necessary so i removed the line

**Comment by kecnry** [comment]
When opening cubeviz, the stretch histogram shows up as an option, but fails to export with `Export failed with: 'NoneType' object has no attribute 'figure'`

**Comment by cshanahan1** [comment]
@kecnry did you pull my latest commit, it fixes that

**Comment by cshanahan1** [comment]
(whoops sorry i pressed close with comment instead of comment)

**Review by kecnry — APPROVED** [review]
LGTM - thanks!

**Comment by haticekaratay** [comment]
This might be only on my end. I see a success message for export  `Exported to cubeviz_export.svg`, but it doesn't show up on the current directory or downloads.

**Comment by cshanahan1** [comment]
Try restarting the kernel? This was happening for me occasionally

**Comment by kecnry** [comment]
This might be susceptible to the same problems as the viewer exporting as its using the same internals.  We'll just have to live what that for now, I think

**Comment by haticekaratay** [comment]
I can't confirm at this point if it works as it should, as clearing the cache, restarting the kernel, and reinstalling the branch didn't solve the problem on my end.

**Comment by cshanahan1** [comment]
@haticekaratay does exporting other items work, is it just the plots that are not working?

**Comment by haticekaratay** [comment]
> @haticekaratay does exporting other items work, is it just the plots that are not working?

Yes, I can export other items. I guess we can merge it as both you can @kecnry confirms that it works. Or another dev might look into it.

**Comment by rosteen** [comment]
I also don't see where the PNG file ends up - I'm trying to rebase on top of #2782 to see if that fixes it, but the rebase is more painful than I expected and I'm out of time tonight. I'll try to continue in the morning.

**Comment by cshanahan1** [comment]
i think i have a hunch about what is going on. What i first noticed was that the stretch histogram is not actually populated with meaningful values until plot_options is opened. If you open the data analysis plugins from the icon not in the viewer, plot_options doesn't open automatically, so the histogram is 'empty'. Once you open that plugin and the histogram is fully created, you should be able to export. I think? this is why I was seeing it working and others weren't - I always opened plot options first. Was there a recent change that stopped this from being opened automatically and thats why we're just noticing it?

The strange thing is, if the histogram were truly 'empty', it wouldn't appear in the export plugin as available for export at all. The filter that is done for plugin plot select checks figure.marks, and in this case it isn't actually empty (there are some initial marks). However, saving it as a png silently fails (even when done from api). It's not until you actually render the figure for display that it is able to be saved as a png.

Therefore, the fact that plot_options isn't opened automatically to render the figure for display is what is causing it to fail when exporting. It isn't technically 'empty', but save_png only works when its been rendered for display (so thats why the viewer being saved to png worked).

**Comment by kecnry** [comment]
This is caused by the same thing that causes the viewers to fail to export until the app is rendered [:cat:](https://jira.stsci.edu/browse/JDAT-3657).  I tried wrapping the call with `plugin.as_active()`, but I think it still needs to be _displayed_ somewhere in order to fully render.  I have a few ideas how to work around this - I'll investigate and report back.

**Comment by rosteen** [comment]
Unfortunately (for that theory) I wasn't trying to save out the plot options histogram, I had run aperture photometry to make a curve of growth plot and was trying to save that, so it was definitely populated.

**Comment by rosteen** [comment]
Also, this might need a rebase to main - I see a bunch of commits from Hatice that I assume aren't supposed to be here. Unless this was building off something else that wasn't merged in addition to Kyle's WIP.

**Comment by kecnry** [comment]
See https://github.com/cshanahan1/jdaviz/pull/1 for a proposed "solution"/workaround.

**Comment by cshanahan1** [comment]
> Also, this might need a rebase to main - I see a bunch of commits from Hatice that I assume aren't supposed to be here. Unless this was building off something else that wasn't merged in addition to Kyle's WIP.

I was working off Kyle's WIP branch which was working off Hatice's WIP branch. I merged main into my branch yesterday too, so maybe that's it. Git history confuses me...

**Comment by cshanahan1** [comment]
> Unfortunately (for that theory) I wasn't trying to save out the plot options histogram, I had run aperture photometry to make a curve of growth plot and was trying to save that, so it was definitely populated.


Ok so I was able to reproduce this, but then refreshing the kernel made the issue go away. Then i refreshed a few more times and it happened again. I can't figure out why it fails only sometimes for me. Does it always fail to export for you?

**Comment by cshanahan1** [comment]
It does seem to happen though consistently when I close the aperture photometry plugin, so hopefully Kyle's PR should fix that

**Comment by rosteen** [comment]
Ahh yes that's it, I confirmed that if I close the aperture photometry plugin the export of the plot fails. If it's open, the save dialog appears as expected.

**Comment by haticekaratay** [comment]
I tested with Cubeviz, and it exports the plots as expected.
image

**Comment by haticekaratay** [comment]
Sorry, I just realized the failing test. Approved to be merged after a fix to that.

**Comment by cshanahan1** [comment]
i'm going to merge this after the tests pass, any objections?

**Pull request merged**